tls/s2n_tls13_handshake.c

```diff
@@ -48,3 +48,118 @@ int s2n_tls13_keys_from_conn(struct s2n_tls13_keys *keys, struct s2n_connection
 
     return 0;
 }
+
+int s2n_tls13_compute_shared_secret(struct s2n_connection *conn, struct s2n_blob *shared_secret)
+{
+    struct s2n_ecc_params *server_key = &conn->secure.server_ecc_params;
+    notnull_check(server_key);
+
+    /* for now we do this tedious loop to find the matching client key selection.
+     * this can be simplified if we get an index or a pointer to a specific key */
+    int selection = -1;
+    for (int i = 0; i < S2N_ECC_SUPPORTED_CURVES_COUNT; i++) {
+        if (server_key->negotiated_curve->iana_id == s2n_ecc_supported_curves[i].iana_id) {
+            selection = i;
+            break;
+        }
+    }
+
+    S2N_ERROR_IF(selection < 0, S2N_ERR_BAD_KEY_SHARE);
+    struct s2n_ecc_params *client_key = &conn->secure.client_ecc_params[selection];
+    notnull_check(client_key);
+
+    if (conn->mode == S2N_CLIENT) {
+        GUARD(s2n_ecc_compute_shared_secret_from_params(client_key, server_key, shared_secret));
+    } else {
+        GUARD(s2n_ecc_compute_shared_secret_from_params(server_key, client_key, shared_secret));
+    }
+
+    return 0;
+}
+
+/*
+ * This function executes after Server Hello is processed
+ * and handshake hashes are computed. It produces and configure
+ * the shared secret, handshake secrets, handshake traffic keys,
+ * and finished keys.
+ */
+int s2n_tls13_handle_handshake_secrets(struct s2n_connection *conn)
+{
+    /* get tls13 key context */
+    s2n_tls13_connection_keys(secrets, conn);
+
+    /* get shared secret */
+    DEFER_CLEANUP(struct s2n_blob shared_secret = { 0 }, s2n_free);
+    GUARD(s2n_tls13_compute_shared_secret(conn, &shared_secret));
+
+    /* derive early secrets */
+    GUARD(s2n_tls13_derive_early_secrets(&secrets));
+
+    /* produce handshake secrets */
+    s2n_stack_blob(client_hs_secret, secrets.size, S2N_TLS13_SECRET_MAX_LEN);
+    s2n_stack_blob(server_hs_secret, secrets.size, S2N_TLS13_SECRET_MAX_LEN);
+
+    struct s2n_hash_state hash_state = {0};
+    GUARD(s2n_handshake_get_hash_state(conn, secrets.hash_algorithm, &hash_state));
+    GUARD(s2n_tls13_derive_handshake_secrets(&secrets, &shared_secret, &hash_state, &client_hs_secret, &server_hs_secret));
+
+    /* produce handshake traffic keys and configure record algorithm */
+    s2n_tls13_key_blob(s_hs_key, conn->secure.cipher_suite->record_alg->cipher->key_material_size);
+    struct s2n_blob s_hs_iv = { .data = conn->secure.server_implicit_iv, .size = S2N_TLS13_FIXED_IV_LEN };
+    GUARD(s2n_tls13_derive_traffic_keys(&secrets, &server_hs_secret, &s_hs_key, &s_hs_iv));
+
+    s2n_tls13_key_blob(c_hs_key, conn->secure.cipher_suite->record_alg->cipher->key_material_size);
+    struct s2n_blob c_hs_iv = { .data = conn->secure.client_implicit_iv, .size = S2N_TLS13_FIXED_IV_LEN };
```

Comment:
preference for client_hs_iv and server_hs_iv to follow the style of client_hs_secret and server_hs_secret, respectively. same with s_app_key & s_app_iv below